### PR TITLE
Reopened: Feature/on request proposal strategy

### PR DESCRIPTION
**Pull request by muratovv**
# Description of the Change
The initial intention of the PR is to introduce a strategy for preventing the creation of proposals in OS. But this pr contains dependent changes for the proper working of the strategy.

# Benefits
Add round optimization for faster collaboration. Also, Closer to release state.

# Possible Drawbacks
A lot of different features which introduce in one PR
List of features:
    * Creation strategy interface in OS
    * Strategy which prevents the creation of proposals
    * OS\OG transport contains peer key in net interaction
    * Supermajority has a new method

# Alternate Designs [optional]
The transport layer is not backward compatible - introduces peer key of the sending side. There is an alternative with meta information in the request.

### CODE REVIEW

File: irohad/ordering/impl/kick_out_proposal_creation_strategy.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      last_requested.insert(*iter);
+    } else {
+      last_requested.insert({peer, RoundType{0, 0}});
+    }
+  }
+  last_requested_ = last_requested;
+}
+
+void KickOutProposalCreationStrategy::shouldCreateRound(
+    RoundType round, const std::function<void()> &on_create) {
```
Comment (igor-egorov) [inline irohad/ordering/impl/kick_out_proposal_creation_strategy.cpp]: It is preferable not to use std::function in non-tests code due to its performance drawbacks.

File: irohad/ordering/impl/on_demand_ordering_gate.cpp [hunk 2]
```diff
@@ -48,7 +50,10 @@ OnDemandOrderingGate::OnDemandOrderingGate(
           round_switch_events.subscribe([this](auto event) {
```
[the remark this answers is not in the document]
Comment (muratovv) [reply irohad/ordering/impl/on_demand_ordering_gate.cpp]: Nice catch!

File: irohad/ordering/impl/kick_out_proposal_creation_strategy.cpp (hunk 1 above)
Comment (muratovv) [reply irohad/ordering/impl/kick_out_proposal_creation_strategy.cpp]: What is your suggestion to use instead?

Comment (muratovv) [reply irohad/ordering/impl/kick_out_proposal_creation_strategy.cpp]: We agreed to leave it as is during an offline discussion.

**Comment by muratovv** [comment]
Closed due to PR is reopened, new PRhttps://github.com/hyperledger/iroha/pull/95.

**Pull request closed without merging**